packages/test/example/src/async.ts

```diff
@@ -0,0 +1,73 @@
+import {
+    Pending,
+    Formula,
+    Primitive,
+    isDelayed,
+    CalcObj,
+    CalcValue,
+    compile
+} from "@tiny-calc/nano";
+
+function createTimeoutPromise<T>(ms: number, value: T, cb: (x: T) => void): Promise<void> {
+    return new Promise(resolve => setTimeout(() => { cb(value); resolve() }, ms));
+}
+
+const cache: Record<string, Primitive | PendingPromise> = {};
+
+interface PendingPromise extends Pending<CalcObj<unknown>> {
+    promise: Promise<void>;
+}
+
+const delayedCalcValue: CalcObj<unknown> = {
+    read(prop) {
+        if (cache[prop] !== undefined) {
+            return cache[prop];
+        }
+
+        let time: number;
+        let val: Primitive;
+
+        switch (prop) {
+            case "a string":
+                time = 100;
+                val = "hello world";
+                break;
+            case "a bool":
+                time = 400;
+                val = true;
+                break;
+            default:
+                time = 600;
+                val = 42;
+                break;
+        }
+        const pending: PendingPromise = {
+            kind: "Pending",
+            promise: createTimeoutPromise(time, val, x => cache[prop] = x)
+        }
+        return cache[prop] = pending;
+    }
+}
+
+const f = compile("{a number} + 10 + {a number} + \" \" + {a string} + \" \" + {a bool}");
+
+function runFormula(f: Formula, attempts: number): Promise<CalcValue<undefined>> {
```

Comment:
I think these kinds of promise loops leak memory. Not an issue here because the chain is short.

Reply:
Confirmed that it does leak on V8.  Might be worth a comment.

That's too bad.  Internally, the runtime could automatically "shorten" the chain by updating the intermediate Promises to point to the end of the chain, thus freeing the intermediate Promises for GC.